draft-pauly-quic-datagram.md

```diff
@@ -140,36 +132,38 @@ LEN bit (0x01). It indicates that there is a Length field present. If this
 bit is set to 0, the Length field is absent and the Datagram Data field extends
 to the end of the packet. If this bit is set to 1, the Length field is present.
 The second least significant bit of the DATAGRAM frame type is the
-DATAGRAM_ID bit (0x02). It indicates that there is a Datagram ID field present.
-If this bit is set to 0, the Datagram ID field is absent and the Datagram ID is
-assumed to be zero. If this bit is set to 1, the Datagram ID field is present.
+FLOW_ID bit (0x02). It indicates that there is a Flow ID field present.
+If this bit is set to 0, the Flow ID field is absent and the Flow ID is
+assumed to be zero. If this bit is set to 1, the Flow ID field is present.
 
-A DATAGRAM frame is shown below.
+The DATAGRAM frame is structured as follows:
 
 ~~~
-0                   1                   2                   3
-0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                      [Datagram ID (i)]                      ...
+|                        [Flow ID (i)]                        ...
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                         [Length (i)]                        ...
+|                        [Length (i)]                         ...
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                       Datagram Data (*)                     ...
+|                      Datagram Data (*)                      ...
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 ~~~
 {: #datagram-format title="DATAGRAM Frame Format"}
 
-The fields of a DATAGRAM frame are as follows:
+DATAGRAM frames contain the following fields:
 
-Datagram ID:
+Flow ID:
 
-: A variable-length integer indicating the datagram ID of the datagram (see
-{{datagram-id}}).
+: A variable-length integer indicating the Flow ID of the datagram (see
+{{flow-id}}). This field is present when the FLOW_ID bit is set, and is assumed
+to be zero otherwise.
 
 Length:
 
-: A variable-length integer specifying the length of the datagram in bytes. If the length
-is zero, the data extends to the end of the QUIC packet.
+: A variable-length integer specifying the length of the datagram in bytes. This field
+is present when the LEN bit is set, and is assumed to be zero otherwise.
```

Comment:
It's not assumed to be zero otherwise, right?  If the LEN bit is not present, it just extends to the end of the QUIC packet.

Reply:
Indeed. The text was trying to get at that by saying:
- if no LEN bit, length == 0
- if length == 0, extends to end

But I understand that that is less clear. Reordered things to say that:
- if LEN bit, length is valid
- if length == 0 or LEN is not present, extends to end

Reply:
I don't think that's correct. We should match STREAM frames in what the length field means. length == 0 means this is a DATAGRAM with no payload, and that's a reasonable frame to send. (Just like sending a UDP datagram with length 0).

Reply:
Ah whoops, you're right. Fixed